### PR TITLE
Adding PFR and StoichiometricReactor Notebook Examples

### PR DESCRIPTION
**Pull request by bpaul4**
## Proposed changes:
- New notebook examples for PFR and StoichiometricReactor unit models and associated index updates

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: notebook_index.yml
Comment (andrewlee94) [inline notebook_index.yml:122]: Not sure what is meant by "pure" liquid phase here. Do you mean "single"?

Also line below.

File: src/notebook_index.ipynb
Comment (andrewlee94) [inline src/notebook_index.ipynb:121]: Same comment r.e. "pure" liquid phase.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:8]: Whilst most readers will probably understand, you should write Plug Flow Reactor out in full the first time you use it.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:25]: Similarly, we probably should expand CSTR the first time. It would also be nice to embed a link to the CSTR example, but I am not sure how easy that is to do with the way our docs get built (@dangunter - any ideas?).

Typo: spatially.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:105]: Should write thermophysical property out in full, package should be plural.

Same for all instances of thermo below - this is a formal document so we should avoid shorthands.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:125]: I am not sure about referencing HELMET and RIPE right now - the code is an absolute mess, and is not tested so we don't even know if they actually work. Both of those are likely to be rebuilt next year, or pushed of to the side somewhere,

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:166]: Again, it would be nice to link to Module 1 here if possible.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:224]: This is not a good idea - a PFT with one element is just a CSTR. It doesn't cost us much to use more elements, so I would suggest we use a few (5 to 10 maybe).

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:225]: I don't think we should include the `length_domain_set` in this example - it is an advanced option and you need to be careful when using it.

Make sure you update any text above if you change these.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:296]: Check the formatting here. On Github's renderer, "\\\\$/kW, and the heating utility cost (\\\\$/s) assuming 2.2E-4 " was all in math model (but that might just be Github).

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:499]: We should add an `InitializationError` (a new exception type) to the property package so that if something breaks in the future we will see exceptions hereto let us know. `InitializationErrors` are currently being added to the unit models in a draft PR in idaes-pse.

**Review by andrewlee94 — CHANGES_REQUESTED** [review]
I haven't look at the stoichiometric reactor notebook yet, but I am guessing many of the comments for the PFR notebook will apply there too.

File: notebook_index.yml
Comment (bpaul4) [reply notebook_index.yml:122]: The three rate-reactor examples exhibit liquid phase behavior only (no vapor).

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:296]: The code renders correctly in the notebook, as well as the chemical reaction equation.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:499]: I will take a look at the PR.

**Comment by bpaul4** [comment]
Thank you @andrewlee94 for your comments, I will review and update all three notebooks for any changes.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:25]: Looking at the `notebook_index.ipynb` file, there is a link to the local host file - I will attempt embedding it behind the text, and do the same for Module 1 below if it is successful.

Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:25]: Update: I successfully added an internal link by copying the "localhost" link generated in `notebook_index.ipynb` and editing the example text to [text to display as blue hyperlink](URL the link opens when clicked)

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:125]: Would you recommend that I remove these references from the examples?

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:166]: See earlier comment on embedding links in Markdown.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:224]: The PFR generates a degree of freedom for each finite element, and the initialization returns an infeasible solution when I fix the heat duty of each element to zero and unfix all other reactor specifications. I'm referring to the [test_pfr](https://github.com/IDAES/idaes-pse/blob/main/idaes/generic_models/unit_models/plug_flow_reactor.py) example as I troubleshoot how to satisfy the degrees of freedom.

Comment (andrewlee94) [reply src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:224]: That indicates that there is a fundamental issue with the formulation of the problem then, which needs to be addressed before we merge this.

**Comment by bpaul4** [comment]
@andrewlee94 I've made the suggested changes in all three rate-reactor notebook examples, including adding the two links. Additionally, I reformulated the initial plug-flow reactor problem which solves correctly now.

File: src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:217]: I will show how to set the number of elements here (even if you use the default value) - it is important for users to know how to do this. We should also show how to set the discretization method (both arguments).

Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/plug_flow_reactor_testing.ipynb:217]: I will add these arguments back in.

File: src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb:94]: I do not think this is used either.

File: src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb:110]: For the stoichiometric reactor, we do not use the rate constant at all so I wonder if it would be better to leave most of this out. All that gets used from the reaction package is the definition of the reactions and their stoichiometry.

File: src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb:398]: I wonder if it is worth mentioning that there are aa number of ways the reaction extent could be defined; e.g. conversion or yield.

File: src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb
Comment (andrewlee94) [inline src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb:721]: The mention of reactor volume here seems out of place, as a stoichiometric reactor by its nature does not depend on volume. I am guessing this is left over from the previous examples.

**Review by andrewlee94 — COMMENTED** [review]
I did a quick read through of the stoichiometric reactor example this time and only had a few comments. Most of them are imports I don't think you need, plus a couple of things that don't make much sense for a stoichiometric reactor, but were left over from the CSTR/PFR example.

File: src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb
Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb:398]: Possibly, conversion is typically more convenient since it is independent of the reaction stoichiometry. There are multiple ways to specify yield (mol EG/mol EO, kg EG/kg EO, or volumetric units) while conversion is more straightforward always being on 0-100% for a given basis. I should also emphasize that conversion is chosen with respect to a specific reactant.

**Comment by bpaul4** [comment]
@andrewlee94 I cleaned up the notebooks and updated the PR

File: src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb
Comment (andrewlee94) [reply src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb]: Not quite - conversion also depends on the stoichiometric basis chosen, as it can be specified with respect to any of the reactants as well, and as reactants are not always fed stoichiometrically a 100% conversion of one reactant is not always equivalent to a 100% conversion of the other (although conventionally we  would use the limiting reagent as the basis, it is just as valid to do it another way).

However, my main point here is that we should highlight that IDAES supports any of these approaches and bases for specifying the reaction extent (including directly setting the extent), and that we are going to use conversion of species X for this example.

Comment (bpaul4) [reply src/Examples/UnitModels/Reactors/stoichiometric_reactor_testing.ipynb]: The section has been updated with the details discussed above.

**Comment by bpaul4** [comment]
@MAZamarripa @jghouse88 This PR is awaiting an additional review. Please let me know if you will have time to review this, or if there is another reviewer I may add. Thanks!

**Review by radhakrishnatg — APPROVED** [review]
Looks good to me.

**Review by andrewlee94 — APPROVED** [review]
I only did a quick skim, so I will trust that my prior reviews were sufficient.

**Pull request merged**
